core/pkg/artifacts/saver.go

```diff
@@ -295,10 +329,164 @@ func newUploadTask(fileInfo serverFileResponse, localPath string) *filetransfer.
 		FileKind: filetransfer.RunFileKindArtifact,
 		Type:     filetransfer.UploadTask,
 		Path:     localPath,
-		Name:     fileInfo.Name,
-		Url:      *fileInfo.UploadUrl,
-		Headers:  fileInfo.UploadHeaders,
+		Name:     fileInfo.name,
+		Url:      *fileInfo.uploadUrl,
+		Headers:  fileInfo.uploadHeaders,
+	}
+}
+
+const (
+	S3MinMultiUploadSize = 2 << 30   // 2 GiB, the threshold we've chosen to switch to multipart
+	S3MaxMultiUploadSize = 5 << 40   // 5 TiB, maximum possible object size
+	S3DefaultChunkSize   = 100 << 20 // 1 MiB
+	S3MaxParts           = 10000
+)
+
+func multiPartRequest(path string) ([]gql.UploadPartsInput, error) {
+	fileInfo, err := os.Stat(path)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get file size for path %s: %w", path, err)
+	}
+	fileSize := fileInfo.Size()
+
+	if fileSize < S3MinMultiUploadSize {
+		// We don't need to use multipart for small files.
+		return nil, nil
+	}
+	if fileSize > S3MaxMultiUploadSize {
+		return nil, fmt.Errorf("file size exceeds maximum S3 object size: %v", fileSize)
+	}
+
+	file, err := os.Open(path)
+	if err != nil {
+		return nil, err
+	}
+	defer file.Close()
+
+	partsInfo := []gql.UploadPartsInput{}
+	partNumber := int64(1)
+	buffer := make([]byte, getChunkSize(fileSize))
+	for {
+		bytesRead, err := file.Read(buffer)
+		if err != nil && err != io.EOF {
+			return nil, err
+		}
+		if bytesRead == 0 {
+			break
+		}
+		partsInfo = append(partsInfo, gql.UploadPartsInput{
+			PartNumber: partNumber,
+			HexMD5:     utils.ComputeHexMD5(buffer[:bytesRead]),
+		})
+		partNumber++
+	}
+	return partsInfo, nil
+}
+
+func (as *ArtifactSaver) uploadMultipart(
+	path string,
+	fileInfo serverFileResponse,
+	partData []gql.UploadPartsInput,
+	doneChan chan<- uploadResult,
+) {
+	statInfo, err := os.Stat(path)
+	if err != nil {
+		doneChan <- uploadResult{name: fileInfo.name, err: err}
+		return
+	}
+	chunkSize := getChunkSize(statInfo.Size())
+
+	type partResponse struct {
+		partNumber int64
+		task       *filetransfer.Task
+	}
+
+	wg := sync.WaitGroup{}
+	subChan := make(chan partResponse, len(partData))
+	// TODO: add mid-upload cancel.
+
+	var contentType string
+	for _, h := range fileInfo.uploadHeaders {
+		if strings.HasPrefix(h, "Content-Type:") {
+			contentType = strings.TrimPrefix(h, "Content-Type:")
+			break
+		}
+	}
```

Comment:
this function is already doing quite a lot, I'd factor this out and give it a nice name

```
contentType := getContentType(fileInfo)
if contentType == "" {
    ....
}